### PR TITLE
Migrate from lalrpop-intern to string-cache.

### LINKED ISSUE #158: port to use `string_cache` instead of `lalrpop_intern`
It seems way better:

https://crates.io/crates/string_cache

### PR DESCRIPTION
**Pull request by ahmedcharles**
Fixes #158.

This probably isn't ready the way it is, but I'd like to get some initial feedback on it.

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@ahmedcharles nice! I've been wanting to do this for a while.

**Comment by ahmedcharles** [comment]
I'm still trying to make the tests pass. And going from a `Copy` type to non-`Copy` is really annoying.

File: lalrpop/src/grammar/repr.rs [hunk 1]
```diff
@@ -222,13 +222,13 @@ impl TypeRepr {
                               None => vec![],
                           })
                           .collect(),
-            TypeRepr::Associated { type_parameter, .. } =>
-                vec![TypeParameter::Id(type_parameter)],
-            TypeRepr::Lifetime(l) =>
-                vec![TypeParameter::Lifetime(l)],
+            TypeRepr::Associated { ref type_parameter, .. } =>
+                vec![TypeParameter::Id(type_parameter.clone())],
```
Comment (nikomatsakis) [inline lalrpop/src/grammar/repr.rs]: sad that `Atom` is not copy; but what you gonna' do

File: lalrpop/src/lr1/build/mod.rs [hunk 2]
```diff
@@ -336,7 +336,7 @@ impl LookaheadBuild for Nil {
 
 impl LookaheadBuild for TokenSet {
     fn epsilon_moves<'grammar>(lr: &LR<'grammar, Self>,
-                               nt: NonterminalString,
+                               nt: &NonterminalString,
```
Comment (nikomatsakis) [inline lalrpop/src/lr1/build/mod.rs]: I definitely approve of this pattern; pass atoms by ref on the stack, clone to move into a struct

Comment (ahmedcharles) [reply lalrpop/src/lr1/build/mod.rs]: My approach so far is that if the function ever passes the type by reference, then it accepts a reference. If it only passes the type by value, then it accepts by value.

File: lalrpop/src/grammar/repr.rs (hunk 1 above)
Comment (ahmedcharles) [reply lalrpop/src/grammar/repr.rs]: I think I agree with it not being copy. I'd probably have more conviction if I hadn't just changed all that code. :)

Comment (nikomatsakis) [reply lalrpop/src/grammar/repr.rs]: Well, the reason is that they maintain a reference count. Seems fine.

File: lalrpop/src/lr1/build/mod.rs (hunk 2 above)
Comment (nikomatsakis) [reply lalrpop/src/lr1/build/mod.rs]: OK, that works. For things that are cheap to clone (e.g., Rc and Arc), I typically take a stronger rule:

- All variables on the stack are by reference
- All things in structs are by value

Comment (nikomatsakis) [reply lalrpop/src/lr1/build/mod.rs]: (But either one is fine.)

**Pull request merged**